server/src/database/models/user.js

```diff
@@ -0,0 +1,74 @@
+
+import bcrypt from 'bcrypt';
+import jwt from 'jsonwebtoken';
+import mongoose from 'mongoose';
+
+const { JWT_SECRET } = process.env;
+
+const userSchema = mongoose.Schema({
+  email: {
+    type: String,
+    required: true,
+    unique: true,
+  },
+  username: {
+    type: String,
+    required: true,
+    trim: true,
+    unique: true,
+    minlength: 3,
+  },
+  password: {
+    type: String,
+    required: true,
+    minlength: 6,
+  },
+});
+
+/**
+ * Overides the model's default JSON stringify method
+ * @returns Object with user ID and username
+ */
+userSchema.methods.toJSON = function toJSON() {
+  const user = this;
+  /* eslint-disable-next-line no-underscore-dangle */
+  return { _id: user._id, username: user.username };
+};
+
+/**
+ * Generates new JWT and returns it
+ * @returns JWT
+ */
+userSchema.methods.generateAuthToken = function generateAuthToken() {
+  const user = this;
+  const access = 'auth';
+  const token = jwt.sign(
+    {
+      /* eslint-disable-next-line no-underscore-dangle */
+      _id: user._id,
+      username: user.username,
+      access,
+    },
+    JWT_SECRET,
+  );
+  return token;
+};
+
+/**
+ * Model pre-save hook to hash the password if it is changed
+ */
+userSchema.pre('save', function preSaveCheck(next) {
+  const user = this;
```

Comment:
So as a personal point of view, try to avoid hooks. I am not saying that hooks are a bad pattern, but it leads to hidden errors like it basically obstructs the natural flow of program i.e you write a command `User.save` and you will be thinking that only it will work, but in really the first hook will be called and then save will be performed.
So better option you just create a function that will hash the password for you and then you can use it

Reply:
I did that because at some point we will give the user the option to change their password. So, this just makes for cleaner flow.

Reply:
From that function you can again call the hash pwd function. And you will know that you have called that function instead of hooks which will happen on their own.

Reply:
Okay, got it. I will implement it without the hook.